tests/integration/utils/instance_cache.py

```diff
@@ -0,0 +1,118 @@
+        """
+        Tests that no db query is used when get_instance is called two times.
+        """
+        CustomSlide.objects.create(title='test cs')
+        get_instance('core/customslide', 1)
+
+        with self.assertNumQueries(0):
+            instance = get_instance('core/customslide', 1)
+        self.assertEqual(instance['title'], 'test cs')
+
+    def test_non_existing_instance(self):
+        with self.assertRaises(CustomSlide.DoesNotExist):
+            get_instance('core/customslide', 1)
+
+    def test_with_user_id(self):
+        CustomSlide.objects.create(title='test cs')
+        instance = get_instance('core/customslide', 1, user=1)
+        self.assertEqual(instance['title'], 'test cs')
+
+    def test_with_anonymous_user(self):
+        """
+        Test to get the instance with the permissions of the anonymous user.
+
+        The anonymous user has not the permission to see custom slide instances
+        """
+        CustomSlide.objects.create(title='test cs')
+        instance = get_instance('core/customslide', 1, user=0)
+        self.assertIsNone(instance)
+
+    def test_with_user_object(self):
+        """
+        There should not be any database query if the object is in the cache and
+        get_instance is called with an user object.
+        """
+        CustomSlide.objects.create(title='test cs')
+        get_instance('core/customslide', 1, user=1)
+        admin = User.objects.get(pk=1)
+
+        with self.assertNumQueries(0):
+            instance = get_instance('core/customslide', 1)
+
+
+class TestInstancesGenerator(TestCase):
+    def test_clean_cache(self):
+        """
+        Tests that the instances are retrieved from the database.
+
+        Currently there are 10 queries needed. This can change in the future,
+        but it has to be more then zero.
+        """
+        CustomSlide.objects.create(title='test cs1')
+        CustomSlide.objects.create(title='test cs2')
+        CustomSlide.objects.create(title='test cs3')
+
+        with self.assertNumQueries(10):
+            instance_list = list(instances_generator('core/customslide'))
+        self.assertEqual(len(instance_list), 3)
+
+    def test_with_cache(self):
+        """
+        Tests that no db query is used when get_instance is called two times.
+        """
+        CustomSlide.objects.create(title='test cs1')
+        CustomSlide.objects.create(title='test cs2')
+        CustomSlide.objects.create(title='test cs3')
+        list(instances_generator('core/customslide'))
+
+        with self.assertNumQueries(0):
+            instance_list = list(instances_generator('core/customslide'))
+        self.assertEqual(len(instance_list), 3)
+
+    def test_with_some_objects_in_the_cache(self):
+        CustomSlide.objects.create(title='test cs1')
+        CustomSlide.objects.create(title='test cs2')
+        list(instances_generator('core/customslide'))
+        CustomSlide.objects.create(title='test cs3')
```

Comment:
This test fails but will probably work in production.

CustomSlide.objects.create() sends a signal to the autoupdate system, which updates the cache. So in production the cache has all data. But in the test environment Channels does not work and therefore the cache system does not know about the third CustomSlide object.

One way to solve this is to not cache the list of ids in a collection but always get this list from the database. This will still be a big performance boost because the only db queries there will be are from Model.objects.value_list('id', flat=True), which are very fast.

There are other problems by caching the ids, because there can be more then one thread, that alters the list of ids at the same time. But on the other hand I would love to have a full cache. @normanjaeckel what do you think?